orttraining/orttraining/training_api/module.cc

```diff
@@ -2,12 +2,137 @@
+    std::shared_ptr<onnxruntime::Model> eval_model;
+    ORT_THROW_IF_ERROR(onnxruntime::Model::Load(eval_model_path_or_bytes.value(), eval_model,
+                                                nullptr, env->GetLoggingManager()->DefaultLogger()));
+    GetGraphInputOutputNames(eval_model->MainGraph(), eval_input_names_, eval_output_names_);
+    // TODO:: do validation on eval inputs and outputs: eg order of user inputs, weights
+    eval_sess_ = std::make_unique<onnxruntime::InferenceSession>(so, *env);
+    ORT_THROW_IF_ERROR(eval_sess_->Load(eval_model_path_or_bytes.value()));
+    ORT_THROW_IF_ERROR(eval_sess_->Initialize());
+  }
+  auto& train_sess_state = train_sess_->GetSessionState();
+
+  std::shared_ptr<onnxruntime::Model> train_model;
+  ORT_THROW_IF_ERROR(onnxruntime::Model::Load(train_model_path_or_bytes, train_model,
+                                              nullptr, env->GetLoggingManager()->DefaultLogger()));
+  GetGraphInputOutputNames(train_model->MainGraph(), train_input_names_, train_output_names_);
+
+  std::vector<std::string> param_input_names, grad_input_names, user_input_names;
+  std::string param_name;
+  for (auto input_name : train_input_names_) {
+    auto it = parameters_.find(input_name);
+    if (it != parameters_.end()) {
+      param_input_names.emplace_back(input_name);
+      weights_.push_back(it->second->Data());
+    } else if (GetParamNameFromGradient(input_name, param_name)) {
+      grad_input_names.emplace_back(input_name);
+      // create gradient buffer
+      // assert param_name is valid.
+      auto it = parameters_.find(param_name);
+      if (it != parameters_.end()) {
+        ORT_THROW_IF_ERROR(it->second->AllocateGrad(input_name, train_sess_state));
+        gradients_.push_back(it->second->Gradient());
+      } else {
+        // raise error here.
+      }
+    } else {
+      user_input_names.emplace_back(input_name);
+    }
+  }
+  train_input_names_ = user_input_names;
+  train_input_names_.insert(train_input_names_.end(), param_input_names.begin(), param_input_names.end());
+  train_input_names_.insert(train_input_names_.end(), grad_input_names.begin(), grad_input_names.end());
+}
+
+std::vector<std::shared_ptr<Parameter>> Module::parameters() const {
+  std::vector<std::shared_ptr<Parameter>> params;
+  for (auto& it : parameters_) {
+    params.push_back(it.second);
+  }
+  return params;
+}
+
+Status Module::ResetGrad() {
+  for (auto& it : parameters_) {
+    ORT_ENFORCE(it.second->ResetGrad().IsOK());
+  }
+  return Status::OK();
+}
+
+Status Module::TrainStep(const std::vector<OrtValue>& inputs, std::vector<OrtValue>& outputs) {
+  std::vector<OrtValue> feeds{inputs};
+  feeds.insert(feeds.end(), weights_.begin(), weights_.end());
+  feeds.insert(feeds.end(), gradients_.begin(), gradients_.end());
+  // TODO: consider maintaining this as ortvalue instead of bool
+  OrtValue reset_grad_input;
+  WarpInOrtValue<bool>(lazy_reset_grad_, &reset_grad_input);
```

Comment:
>reset_grad_input [](http://example.com/codeflow?start=42&length=16)

is this always false?

Reply:
no this has to be done as a next step to reset the gradient lazily

Reply:
how the lazy grad reset works?

Reply:
adding the functionality in this PR itself